tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/TaskContinuationGenerator.cs

```diff
@@ -6,93 +6,187 @@
+                    (result.Method.ReturnType.IsGenericType && typeof(Task).IsAssignableFrom(result.Method.ReturnType)))
+                {
+                    var asyncContinuation = (AsyncObjectContinuationMethodDelegate)result.Method.CreateDelegate(typeof(AsyncObjectContinuationMethodDelegate));
+                    Resolver = new AsyncCallbackHandler(asyncContinuation, result.PreserveContext);
+                }
+                else
+                {
+                    var continuation = (ObjectContinuationMethodDelegate)result.Method.CreateDelegate(typeof(ObjectContinuationMethodDelegate));
+                    Resolver = new SyncCallbackHandler(continuation, result.PreserveContext);
+                }
+            }
+            else
+            {
+                Resolver = new NoOpCallbackHandler();
             }
-        }
 
-        internal delegate object ContinuationMethodDelegate(TTarget target, object returnValue, Exception exception, in CallTargetState state);
+            if (Log.IsEnabled(LogEventLevel.Debug))
+            {
+                Log.Debug($"== TaskContinuationGenerator<{typeof(TIntegration).FullName}, {typeof(TTarget).FullName}, {typeof(TReturn).FullName}> using Resolver: {Resolver.GetType().FullName}");
+            }
+        }
 
         public override TReturn SetContinuation(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
         {
-            if (_continuation == null)
+            return Resolver.ExecuteCallback(instance, returnValue, exception, in state);
+        }
+
+        private class SyncCallbackHandler : CallbackHandler
+        {
+            private readonly ObjectContinuationMethodDelegate _continuation;
+            private readonly bool _preserveContext;
+
+            public SyncCallbackHandler(ObjectContinuationMethodDelegate continuation, bool preserveContext)
             {
-                return returnValue;
+                _continuation = continuation;
+                _preserveContext = preserveContext;
             }
 
-            if (exception != null || returnValue == null)
+            public override TReturn ExecuteCallback(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
             {
-                _continuation(instance, default, exception, in state);
-                return returnValue;
+                if (exception != null || returnValue == null)
+                {
+                    _continuation(instance, default, exception, in state);
+                    return returnValue;
+                }
+
+                Task previousTask = FromTReturn<Task>(returnValue);
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    _continuation(instance, default, null, in state);
+                    return returnValue;
+                }
+
+                return ToTReturn(ContinuationAction(previousTask, instance, state));
```

Comment:
You can't add `in` to the `ContinuationAction` method because it's `async`, but you can add it here I think. I'm not sure if it has any value, but probably doesn't do any harm at least! Up to you 🙂 

```suggestion
                return ToTReturn(ContinuationAction(previousTask, instance, in state));
```

Reply:
<img width="309" alt="image" src="https://user-images.githubusercontent.com/69803/208126944-fe4ce925-e02c-4d48-9572-78f559d4e8c0.png">
😢